File: test-examples/sse-chat-subscribe.ts
// npm install eventsource
import { EventSource } from 'eventsource';
import { config } from 'dotenv';
config({ path: '.env.test' });

const baseUrl = process.env.TEST_BASE_URL || 'http://localhost:3010';
const apiKey = process.env.TEST_API_KEY!;
const clientId = process.argv[2] || process.env.TEST_CLIENT_ID_V13 || process.env.TEST_CLIENT_ID_V12;

if (!clientId) {
  console.error('Usage: tsx test-examples/sse-chat-subscribe.ts <clientId>');
  console.error('  Or set TEST_CLIENT_ID_V13 in .env.test');
  process.exit(1);
}

const url = `${baseUrl}/chat/subscribe?clientId=${clientId}`;
console.log(`Connecting to ${url}...`);

// eventsource v4 uses a custom fetch function to inject headers
const eventSource = new EventSource(url, {
  fetch: (input, init) => fetch(input, {
    ...init,
    headers: { ...init?.headers, 'x-api-key': apiKey }
  })
});

interface DieResult {
  result: number;
  active: boolean;
}

interface Die {
  faces: number;
  results: DieResult[];
}

interface Roll {
  formula: string;
  total: number;
  isCritical: boolean;
  isFumble: boolean;
  dice: Die[];
}

interface ChatMessage {
  id: string;
  content: string;
  type: string;
  author: { id: string; name: string };
  timestamp: number;
  flavor: string;
  isRoll: boolean;
  rolls: Roll[];
  speaker: any;
  whisper: string[];
  flags: Record<string, any>;
}

function formatMessage(prefix: string, message: ChatMessage) {
  const speaker = message.author?.name || message.speaker?.alias || '?';
  console.log(`[${prefix}] ${speaker}: ${message.content}`);
  if (message.flavor) console.log(`  Flavor: ${message.flavor}`);
  if (message.isRoll && message.rolls?.length > 0) {
    for (const roll of message.rolls) {
      const dice = roll.dice?.map(d =>
        `${d.results.map(r => `${r.result}${r.active ? '' : '(dropped)'}`).join(', ')} (d${d.faces})`
      ).join(' + ') || '';
      console.log(`  Roll: ${roll.formula} = ${roll.total}${roll.isCritical ? ' CRITICAL!' : ''}${roll.isFumble ? ' FUMBLE!' : ''}`);
      if (dice) console.log(`  Dice: ${dice}`);
    }
  }
}

eventSource.addEventListener('connected', (event: MessageEvent) => {
  const data = JSON.parse(event.data);
  console.log('Connected:', data.clientId);
  console.log('Listening for chat events... (Ctrl+C to stop)\n');
});

eventSource.addEventListener('chat-create', (event: MessageEvent) => {
  const message: ChatMessage = JSON.parse(event.data);
  formatMessage('new', message);
});

eventSource.addEventListener('chat-update', (event: MessageEvent) => {
  const message: ChatMessage = JSON.parse(event.data);
  formatMessage('updated', message);
});

eventSource.addEventListener('chat-delete', (event: MessageEvent) => {
  const data = JSON.parse(event.data);
  console.log('Message deleted:', JSON.stringify(data));
});

eventSource.onerror = (error) => {
  console.error('SSE error:', error);
  eventSource.close();
  process.exit(1);
};

// Graceful shutdown
process.on('SIGINT', () => {
  console.log('\nDisconnecting...');
  eventSource.close();
  process.exit(0);
});

File: test-examples/sse-chat-subscribe.py
#!/usr/bin/env python3
"""
SSE Chat Subscribe example
pip install sseclient-py requests python-dotenv

Usage: python test-examples/sse-chat-subscribe.py <clientId>
  Or set TEST_CLIENT_ID_V13 in .env.test
"""

import sys
import os
import json

try:
    import sseclient
    import requests
    from dotenv import load_dotenv
except ImportError as e:
    print(f"Missing dependency: {e}")
    print("Install with: pip install sseclient-py requests python-dotenv")
    sys.exit(1)

load_dotenv('.env.test')

base_url = os.environ.get('TEST_BASE_URL', 'http://localhost:3010')
api_key = os.environ.get('TEST_API_KEY', '')
client_id = sys.argv[1] if len(sys.argv) > 1 else os.environ.get('TEST_CLIENT_ID_V13') or os.environ.get('TEST_CLIENT_ID_V12')

if not client_id:
    print('Usage: python test-examples/sse-chat-subscribe.py <clientId>')
    print('  Or set TEST_CLIENT_ID_V13 in .env.test')
    sys.exit(1)

url = f'{base_url}/chat/subscribe'
params = {'clientId': client_id}
headers = {
    'x-api-key': api_key,
    'Accept': 'text/event-stream'
}

print(f'Connecting to {url}?clientId={client_id}...')

response = requests.get(url, params=params, headers=headers, stream=True)

if response.status_code != 200:
    print(f'Error: HTTP {response.status_code}')
    print(response.text)
    sys.exit(1)

client = sseclient.SSEClient(response)

try:
    for event in client.events():
        data = json.loads(event.data)

        if event.event == 'connected':
            print(f'Connected: {data["clientId"]}')
            print('Listening for chat events... (Ctrl+C to stop)\n')
        elif event.event in ('chat-create', 'chat-update'):
            prefix = 'new' if event.event == 'chat-create' else 'updated'
            speaker = (data.get('author') or {}).get('name') or (data.get('speaker') or {}).get('alias') or '?'
            print(f'[{prefix}] {speaker}: {data.get("content", "")}')
            if data.get('flavor'):
                print(f'  Flavor: {data["flavor"]}')
            if data.get('isRoll') and data.get('rolls'):
                for roll in data['rolls']:
                    dice_parts = []
                    for d in roll.get('dice', []):
                        results = ', '.join(
                            f'{r["result"]}{"" if r.get("active", True) else "(dropped)"}'
                            for r in d.get('results', [])
                        )
                        dice_parts.append(f'{results} (d{d["faces"]})')
                    crit = ' CRITICAL!' if roll.get('isCritical') else ''
                    fumble = ' FUMBLE!' if roll.get('isFumble') else ''
                    print(f'  Roll: {roll["formula"]} = {roll["total"]}{crit}{fumble}')
                    if dice_parts:
                        print(f'  Dice: {" + ".join(dice_parts)}')
        elif event.event == 'chat-delete':
            print(f'Message deleted: {json.dumps(data)}')
        else:
            print(f'Unknown event: {event.event} -> {data}')
except KeyboardInterrupt:
    print('\nDisconnecting...')
